Add multi-cluster support (showMultiCluster=True) to alerts

frittentheke opened the pull request:
Following PR https://github.com/ceph/ceph/pull/55495 fixing the dashboard in regards to multiple clusters storing their metrics in a single Prometheus instance, this PR addresses the issues for alerts.

Fixes: https://tracker.ceph.com/issues/64321





<!--
  - Please give your pull request a title like

      [component]: [short description]

  - Please use this format for each git commit message:

      [component]: [short description]

      [A longer multiline description]

      Fixes: [ticket URL on tracker.ceph.com, create one if necessary]
      Signed-off-by: [Your Name] <[your email]>

    For examples, use "git log".
-->

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "quincy"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

- When filling out the below checklist, you may click boxes directly in the GitHub web UI.  When entering or editing the entire PR message in the GitHub web UI editor, you may also select a checklist item by adding an `x` between the brackets: `[x]`.  Spaces and capitalization matter when checking off items this way.

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [ ] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [ ] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [ ] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
- `jenkins test rook e2e`
</details>

frittentheke commented:
@cloudbehl this here is the the split off from PR https://github.com/ceph/ceph/pull/55495 addressing only the alerts
It's not done yet (see comment https://github.com/ceph/ceph/pull/55495#issuecomment-2043714581). But I shall push another revision when I am done. I'd really appreciate some feedback and maybe even some hints on how to address the remaining PromQL queries properly.

frittentheke commented:
@cloudbehl with https://github.com/ceph/ceph/pull/55495 merged now, should we tackle the alerts and make them multiCluster compatible at some point?

cloudbehl commented:
@frittentheke yes lets do it. Let me know once this is ready for review and I can start the testing.

frittentheke commented:
> @frittentheke yes lets do it. Let me know once this is ready for review and I can start the testing.

@cloudbehl sorry for the delay. I pushed through all those failing tests and alerts now.
PTA (good) L.

frittentheke commented:
@cloudbehl I just rebased this again (and fixed the tox allowlist regarding the call of `./lint-jsonnet.sh`).
Did you get a chance to review this? 

I'd really love to get this done soon, as I am leaving the field of running Ceph in production myself and it becomes a little harder for me to keep refining this one. I gladly fix and adjust things you find during review, but just wanted to be transparent.

cloudbehl commented:
> @cloudbehl I just rebased this again (and fixed the tox allowlist regarding the call of `./lint-jsonnet.sh`). Did you get a chance to review this?
> 
> I'd really love to get this done soon, as I am leaving the field of running Ceph in production myself and it becomes a little harder for me to keep refining this one. I gladly fix and adjust things you find during review, but just wanted to be transparent.

Sorry, I was busy with other things. I will test and update you tomorrow related to same.

cloudbehl commented:
@frittentheke I tested the basic feature with some alerts and its working fine. I think I want to do another round of  testing if something doesn't work we might end up breaking whole monitoring system.

cloudbehl reviewed (approved):
LGTM

frittentheke commented:
> @frittentheke I tested the basic feature with some alerts and its working fine. I think I want to do another round of testing if something doesn't work we might end up breaking whole monitoring system.

@cloudbehl let me know if there is anything left for me to do!

aaSharma14 commented:
Hi @frittentheke , Thank you for the PR.

I was testing this PR and found an issue -

![Screenshot from 2024-08-06 16-33-11](https://github.com/user-attachments/assets/0a409518-09a9-4393-ac91-2248dce934d7)

The issue is in the description of the `CephOSDDownHigh` alert - ` error calling query: found duplicate series for the match group {ceph_daemon="osd.0"} on the right hand-side of the operation: `. The issue is because different clusters can have same `ceph_daemon` name. I was able to get rid of the error by modifying the query used in the `CephOSDDownHigh` alert's description to `ceph_osd_up * on(ceph_daemon, cluster) group_left(hostname) ceph_osd_metadata == 0` (basically adding `cluster` label to `on(ceph_daemon)`). I found some similar implementations here in these alerts as well -

CephMonDownQuorumAtRisk - description
CephMonDown - description
CephOSDDownHigh - description
CephOSDHostDown - description
CephOSDDown - description
CephNodeDiskspaceWarning - expr
CephPoolFull - description
NVMeoFHighHostCPU - expr

frittentheke commented:
@aaSharma14  @cloudbehl  sorry for the delay again.

I now reworked the description (and also aligned the use of  jsonnet multiline (`|||`) vs. simple strings `''` and cleaned up some (whitespace, newline) noise in the expected descriptions.


PTAL.

frittentheke commented:
@aaSharma14 @cloudbehl did you get the time to take a look at my recent update to this PR?

aaSharma14 commented:
> @aaSharma14 @cloudbehl did you get the time to take a look at my recent update to this PR?

Thanks again for the update @frittentheke , i will test it again and then probably we can merge this PR.

aaSharma14 commented:
jenkins test make check

aaSharma14 commented:
jenkins test make check

cloudbehl commented:
jenkins test make check

cloudbehl commented:
I see some unit test failures

```
Unit Testing:  /home/jenkins-build/build/workspace/ceph-pull-requests/monitoring/ceph-mixin/tests_alerts/test_alerts.yml
  FAILED:
    expr: "(\n  rate(ceph_osd_up[5m])\n  * on(ceph_daemon) group_left(hostname) ceph_osd_metadata\n) * 60 > 1\n", time: 1m,
        exp:"{ceph_daemon=\"osd.0\", cluster=\"mycluster\", hostname=\"ceph\", instance=\"ceph:9283\", job=\"ceph\"} 1.21E+01"
        got:"{ceph_daemon=\"osd.0\", cluster=\"mycluster\", hostname=\"ceph\", instance=\"ceph:9283\", job=\"ceph\"} 1.2200000000000001E+01"
        ```

aaSharma14 commented:
jenkins test make check

aaSharma14 commented on monitoring/ceph-mixin/tests_alerts/test_alerts.yml:
```diff
@@ -142,8 +145,8 @@ tests:
        exp_samples:
          - labels: '{ceph_daemon="osd.0", hostname="ceph", instance="ceph:9283",
-           job="ceph"}'
-           value: 1.2200000000000001E+01
```
@frittentheke , I think the error in make-check is because of this change. Can you please try reverting this change and check if the tests pass

aaSharma14 replied:
@frittentheke ^^

aaSharma14 reviewed (approved):
Tested Locally, Works fine. Thanks @frittentheke for the contribution.

aaSharma14 commented:
jenkins test make check

aaSharma14 commented:
jenkins test api

aaSharma14 commented:
jenkins test make check

cloudbehl commented:
jenkins test make check
